includes/ThanksCache.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace MediaWiki\Extension\Thanks;
+
+use RequestContext;
+use Wikimedia\Rdbms\ILoadBalancer;
+
+/**
+ * A service class that caches the list of IDs (revisions, logs) a user has thanked for.
+ * It uses a session as a backend.
+ */
+class ThanksCache {
+	private const THANKED_IDS_SESSION_KEY_PATTERN = "thanks-thanked-ids-%d";
+
+	public function __construct( private ILoadBalancer $lb ) {
+	}
+
+	/**
+	 * Gets a cached list of entires for which a user has thanked.
+	 * Entires are refetched on cache miss.
+	 *
+	 * Each entry is in the same format as saved by the store (i.e. a string formatted as "$type-$id").
+	 * @param int $thankerActorId - actor ID of the user who thanked.
+	 * @return array An array of strings formatted as "$type-$id"
+	 */
+	public function getUserThanks( int $thankerActorId ): array {
+		$dbr = $this->lb->getConnectionRef( DB_REPLICA );
+		$session = RequestContext::getMain()->getRequest()->getSession();
+
+		$key = sprintf( self::THANKED_IDS_SESSION_KEY_PATTERN, $thankerActorId );
+		$cachedRevs = $session->get( $key );
+		if ( $cachedRevs !== null ) {
+			return $cachedRevs;
+		}
+
+		$dbEntries = $dbr->newSelectQueryBuilder()
+			->select( 'ls_value' )
+			->from( 'logging' )
+			->join( 'log_search', conds: [ 'log_id = ls_log_id' ] )
+			->where(
+				[
+					'log_actor' => $thankerActorId,
+					'log_type' => 'thanks',
+					'ls_field' => 'thankid',
```

Comment:
issue: This appears to be an unbounded query (it will be more and more difficult to compute, once the user submitted enough thanks). I'm slightly worried about its performance for users with many (tm) thanks on their file. 

suggestion: AFAICS, getUserThanks() is only used within ThanksCache here. Can we make a query that checks for a single ls_value, instead of fetching all values and checking that list (see LogStore::haveThanked).

Reply:
In theory we might filter by time, and only look for thanks that are close to now. I imagine double thanking for the same edit across 10 years wouldn't seem weird (in fact, it might even be intended). I'm not sure what the right cutoff would be, or if it would be of help. 

I tried to run the query in WMF's environment (for English Wikipedia's largest thanker), and it took about 10 seconds. For frontend facing stuff, this is definitely a lot.

Reply:
Allright, I've introduced a configurable limit to the amount of days the query should look back. And yes, it may still potentially take quite some time to execute, but its going to be cached for next requests, so I wouldn't worry too much about the initial load.